merlin/models/tf/models/base.py

```diff
@@ -1408,4 +1408,7 @@ def _maybe_convert_merlin_dataset(data, batch_size, shuffle=True, **kwargs):
         if not shuffle:
             kwargs.pop("shuffle", None)
 
+    if isinstance(data, Loader):
+        data = (batch for batch in data)
```

Comment:
Converting `Loader` into a generator here to ensure it works correctly with the Keras [GeneratorDataAdapter](https://github.com/keras-team/keras/blob/v2.10.0/keras/engine/data_adapter.py#L828). When passing the `Loader` to `GeneratorDataAdapter` it results in incorrect number of batches due to this [peek_and_restore](https://github.com/keras-team/keras/blob/v2.10.0/keras/engine/data_adapter.py#L933-L936) method

Reply:
An alternative or additional option could be to update the `Dataloader` class to be more like an iterator in the way the `next`/`StopIteration` work. 
 
Currently, the `Loader` after raising `StopIteration` once, will start to cycle through the batches again. Whereas a regular python generator will continue to raise `StopIteration` each time `next` is called on a generator with no items left.